### ISSUE TITLE
XSD with targetNamespace cannot be used with xml.fileAssociations

### ISSUE THREAD

**Issue by 747** [redhat-developer/vscode-xml#223]
I have an originally error-free file:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<TEI xmlns="http://www.tei-c.org/ns/1.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.tei-c.org/ns/1.0 https://tei-c.org/Vault/P5/current/xml/tei/custom/schema/xsd/tei_all.xsd">
  <teiHeader>
    <fileDesc>
      <titleStmt><title>dummy</title></titleStmt>
      <publicationStmt><publisher>dummy</publisher></publicationStmt>
      <sourceDesc><p>dummy</p></sourceDesc>
    </fileDesc>
  </teiHeader>
  <text>
    <body><p>dummy</p></body>
  </text>
</TEI>
```
When I tried to replace it with `xml.fileAssociations`:
```
"xml.fileAssociations": [
        {
            "systemId": "https://tei-c.org/Vault/P5/current/xml/tei/custom/schema/xsd/tei_all.xsd",
            "pattern": "**/*.xml"
        },
        {
            "systemId": "http://www.w3.org/2001/XMLSchema.xsd",
            "pattern": "**/*.xsd"
        }
    ]
```
and changed the line 2 as:
```xml
<TEI xmlns="http://www.tei-c.org/ns/1.0">
```
I got this error:
```
cvc-elt.1.a: Cannot find the declaration of element 'TEI'.
```

---
I've found a similar case in #215, but the answer given there doesn't seem to answer me because what I want isn't replace the remote schema with a local file. What am I misunderstanding on the behavior of `xml.fileAssociations`?

**Comment by angelozerr** [comment]
> I've found a similar case in #215, but the answer given there doesn't seem to answer me because what I want isn't replace the remote schema with a local file. 

With XML catalog, your XSD schema is in your local file. Why you could not use XML catalog?

**Comment by 747** [comment]
@angelozerr Sorry if have a big misconception about the use case of `xml.fileAssociations`. And if you can confirm that using `xml.catalogs` is the only means in this library, I'm happy to do it. But I have questions which so far I couldn't find any documentation for.

1. What is the typical usage of `xml.fileAssociations`? From the Wiki example I assumed that this was my primary choice which can globally replace the per-file `schemaLocation` declaration. Also why does the error happen in this way?
2. My original intention was, as you can see, referring to a URL that always points to the latest standard schema, and that's what XML catalog doesn't cover. Is my practice prohibited, discouraged, or not recommended in some way? Isn't it a part of the purpose of `xml.fileAssociations` is provided alongside `xml.catalogs`?

**Comment by a2975667** [comment]
I believe I am having similar issues. Are there any follow-ups?

**Comment by angelozerr** [comment]
Sorry @747 I tried your sample and it works great for me without file association. You don't need to use file association and XML catalog (like I suggested) because you are using xsi:schemaLocation. vscode-xml should download the declared XSD and after that it should work (I tested with last version of vscode-xml).

Please retry without xml fileassociations declaration, if it doesn't work, try to use last version of vscode-xml, if it doesn't work, do you some error about downloading (in the TEI root element). Are you using proxy?

@a2975667 this error comes from when XSD is not proper referenced, I suggest you that you share your sample.

**Comment by 747** [comment]
I should have explained more about the background information.

What I originally wanted to do was to replace tooltip's language without touching the original XML. There are many localized versions of the [TEI](https://tei-c.org/) schema, but only the official one (in English) has a proper URI to refer to. Others can be generated with [Roma](https://roma.tei-c.org/) to be used locally. They work same as each other except for the doc language, so I'm seeking a way to let coders see description in their language without hardcoding its path in XML files.

In my original post, I put the same remote URI in `systemId` just because I wanted to make sure that it was not my fault having wrong path or anything, but even the same URI didn't work as when it was in `xsi:schemaLocation`. Though I'm still curious about that, my objective was to associate a custom local XSD with XMLs.

What? Are you messing around XML when you don't read English!? I mean, TEI mostly serves for humanities researchers, where each different locale has its own history and philological tradition and whatnot, that makes description in English terminology not necessarily helpful. I don't think it's a very eccentric idea in this field...

**Comment by angelozerr** [comment]
xml.fileAssociations is a non standard mean to bind a XML with a XSD grammar without declaring the association in the XML. But it doesn't take care from namespace, so in your case you cannot use it.

XML catalog gives the capability to define XSD in local insead of using remote XSD (declared with XSD) which uses namespaces. So I think XML catalog is for your usecase. 

But to be honnest with you I don't understand very well what you want to do. I suggest that you attach a zip with xml, xsd that you wish to use.

**Comment by 747** [comment]
> But it doesn't take care from namespace, so in your case you cannot use it.

So this is the root cause. Is there any chance that this restriction will be lifted? If not in near future, you might want to warn user about it in Wiki page or somewhere, as it doesn't seem a self-evident behavior.

I myself am using XML Catalogs now, so this issue can be closed whenever the measure taken.

**Comment by a2975667** [comment]
@angelozerr Here is my setup.
In short, I want to develop a plugin that comes with xsds that I can update form time to time and supports a specific filename so other people do not need to specify the uri for the schema.

**file structure**
```
--Sample
---- .plugin/resource
-------- sample.xsd
---- .vscode
-------- settings.json
---- project_folder
--------target-file.xml
```

sample.xsd
```xsd
<xs:schema attributeFormDefault="unqualified" elementFormDefault="qualified" targetNamespace="http://some.address.com/2006/04/metadata" xmlns:xs="http://www.w3.org/2001/XMLSchema">
  <xs:element name="LayerOne" type="LayerOneType" xmlns="http://some.address.com/2006/04/metadata"/>

  <xs:complexType name="LayerOneType">
    <xs:sequence>
      <xs:element type="xs:double" name="element"/>
    </xs:sequence>
  </xs:complexType>

</xs:schema>
```

settings.json
```json
 { 
  "xml.fileAssociations": [
    {
        "systemId": ".plugin/resources/sample.xsd", 
        "pattern": "**/*target-file.xml"
    }
  ]
 }
```

target-file.xml
```xml
<?xml version="1.0" encoding="UTF-8"?>
<LayerOne xmlns="http://some.address.com/2006/04/metadata" >
    <element>22</element>
</LayerOne>
```

In this case, LayerOne will have the following error:
`cvc-elt.1.a: Cannot find the declaration of element 'LayerOne'.xml(cvc-elt.1.a)`

**Comment by angelozerr** [comment]
As I explained if you use namespace you cannot use file association. Please use XML catalog.

**Comment by a2975667** [comment]
Thanks for confirmation! :)

**Comment by angelozerr** [comment]
> As I explained if you use namespace you cannot use file association. Please use XML catalog.

@a2975667 I'm implementing this issue which should be available soon.

> So this is the root cause. Is there any chance that this restriction will be lifted?

@747 I'm implementing this issue which should be available soon.

**Comment by datho7561** [comment]
Closed by #357
